src/read/xz.rs

```diff
@@ -0,0 +1,330 @@
+use crc32fast::Hasher;
+use lzma_rs::decompress::raw::Lzma2Decoder;
+use std::{
+    collections::VecDeque,
+    io::{BufRead, BufReader, Error, Read, Result, Write},
+};
+
+#[derive(Debug)]
+pub struct XzDecoder<R> {
+    compressed_reader: BufReader<R>,
+    stream_size: usize,
+    buf: VecDeque<u8>,
+    check_size: usize,
+    records: Vec<(usize, usize)>,
+    flags: [u8; 2],
+}
+
+impl<R: Read> XzDecoder<R> {
+    pub fn new(inner: R) -> Self {
+        XzDecoder {
+            compressed_reader: BufReader::new(inner),
+            stream_size: 0,
+            buf: VecDeque::new(),
+            check_size: 0,
+            records: vec![],
+            flags: [0, 0],
+        }
+    }
+}
+
+struct CountReader<'a, R: BufRead> {
+    inner: &'a mut R,
+    count: &'a mut usize,
+}
+
+impl<R: BufRead> Read for CountReader<'_, R> {
+    fn read(&mut self, buf: &mut [u8]) -> Result<usize> {
+        let count = self.inner.read(buf)?;
+        *self.count += count;
+        Ok(count)
+    }
+}
+
+impl<R: BufRead> BufRead for CountReader<'_, R> {
+    fn fill_buf(&mut self) -> Result<&[u8]> {
+        self.inner.fill_buf()
+    }
+
+    fn consume(&mut self, amt: usize) {
+        self.inner.consume(amt);
+        *self.count += amt;
+    }
+}
+
+struct BufWriter<'a> {
+    inner: &'a mut [u8],
+    written: &'a mut usize,
+    total: &'a mut usize,
+    rest: &'a mut VecDeque<u8>,
+}
+
+impl<'a> Write for BufWriter<'a> {
+    fn write(&mut self, buf: &[u8]) -> Result<usize> {
+        if self.inner.len() > *self.written {
+            let len = std::cmp::min(buf.len(), self.inner.len() - *self.written);
+            self.inner[*self.written..*self.written + len].copy_from_slice(&buf[..len]);
+            *self.written += len;
+            *self.total += buf.len();
+            Ok(len)
+        } else {
+            self.rest.extend(buf.iter().copied());
```

Comment:
How does `self.rest` ever get cleared?

Reply:
see line 106. it will be consumed by next call to read.